File: code/ch02/2.6.3.maxSubArray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        n = len(nums)
        max_sum_ending_curr_index = max_sum = nums[0]
        for i in range(1, n):
            max_sum_ending_curr_index = max(
                max_sum_ending_curr_index + nums[i], nums[i]
            )
            max_sum = max(max_sum_ending_curr_index, max_sum)

        return max_sum

File: code/ch02/2.6.3.maxSubArray.java
import java.lang.Math;

public class Solution {
  public int maxSubArray(int[] nums) {
    int n = nums.length;
    int max_sum_ending_curr_idx = nums[0];
    int max_sum = nums[0];
    for (int i = 1; i < n; i++) {
      max_sum_ending_curr_idx = Math.max(
          max_sum_ending_curr_idx + nums[i],
          nums[i]
          );
      max_sum = Math.max(
          max_sum_ending_curr_idx,
          max_sum
          );
    }
    return max_sum;
  }
}
